src/mixins/java/org/spongepowered/common/mixin/core/entity/player/ServerPlayerEntityMixin.java

```diff
@@ -468,26 +495,144 @@ public void teleportTo(net.minecraft.world.server.ServerWorld world, double x, d
+                // Sponge End
+                if (portalinfo != null) {
+                    // Sponge Start: wrap the teleportation logic within a function to allow for modification
+                    // of the teleporter, as well as calling events(?)
+                    final Vector3d originalDestination = new Vector3d(portalinfo.pos.x, portalinfo.pos.y, portalinfo.pos.z);
+                    final Entity e = spongeTeleporter.performTeleport((Entity) (Object) this, serverworld, targetWorld, this.xRot, spawnPortal -> {
+                        // Sponge End
+                        serverworld.getProfiler().push("moving");
+                        if (registrykey == World.OVERWORLD && targetWorld.dimension() == World.NETHER) {
+                            this.enteredNetherPosition = this.shadow$position();
+                        // Sponge Start: only enter this branch if we're going through portal logic
+                        } else if (spawnPortal && targetWorld.dimension() == World.END) {
+                            // Sponge End
+                            this.shadow$createEndPlatform(targetWorld, new BlockPos(portalinfo.pos));
+                        }
+
+                        serverworld.getProfiler().pop();
+                        serverworld.getProfiler().push("placing");
+                        this.shadow$setLevel(targetWorld);
+                        targetWorld.addDuringPortalTeleport((ServerPlayerEntity) (Object) this);
+                        this.shadow$setRot(portalinfo.yRot, portalinfo.xRot);
+                        this.shadow$moveTo(portalinfo.pos.x, portalinfo.pos.y, portalinfo.pos.z);
+                        serverworld.getProfiler().pop();
+                        // Sponge Start: return this entity
+                        return (Entity) (Object) this;
+                    });
+                    // Sponge End
+                    // This is done here (it's not in Forge) to ensure that it's done. This is to ensure
+                    // advancements are triggered.
+                    this.shadow$triggerDimensionChangeTriggers(serverworld);
+                    // Sponge Start: make sure the right object was returned
+                    if (e != (Object) this) {
+                        throw new IllegalArgumentException(String.format("Teleporter %s "
+                                + "did not return the expected player entity: got %s, expected PlayerEntity %s", spongeTeleporter, e, this));
+                    }
+
+                    // If we need to reposition: well... reposition.
+                    // Downside: portals won't come with us, but with how it's implemented in Forge,
+                    // not sure how we'd do this
+                    final Vector3d destination = VecHelper.toVector3d(this.shadow$position());
+                    final ChangeEntityWorldEvent.Reposition reposition = SpongeEventFactory.createChangeEntityWorldEventReposition(
```

Comment:
Here is the big difference between my version of this and yours.

In my version, I run reposition logic by finding where the portal will put you but I don't place any portal blocks. If the event succeeds, I run the logic again based on the event's position and put the blocks there. I did this to prevent cancelling reposition would end up having to roll back blocks in the PhaseTracker.

Reply:
This is the big problem I have.

For vanilla, that's okay. We can do some magic with `getExitPortal` (or just copy/modify a redirection of it), and that's what my one of my TODOs is about. So if this is a route we wanted to go down, I would have spent some time doing that for vanilla logic - this wouldn't remain here if vanilla logic is running.

For Forge, there's very little we can do because there's no way to feed in the requested position to the mod.

Reply:
You are correct, in my version if a mod passed in their own lambda we would call no events. It was definitely a flaw but I accepted it cause the whole thing is asinine to begin with.............